File: src/main/java/cspfj/constraint/extension/MatrixManager2D.java
package cspfj.constraint.extension;

import cspfj.problem.Variable;
import cspfj.util.BitVector;

public final class MatrixManager2D extends AbstractMatrixManager {

    private static final int MINIMUM_SIZE_FOR_LAST = 3 * Long.SIZE;

    private Matrix2D matrix;

    private final int[][] last;

    private static long checks = 0;

    private static long presenceChecks = 0;

    public MatrixManager2D(final Variable[] scope, final Matrix2D matrix,
            final boolean shared, final int[] tuple) {
        super(scope, matrix, shared, tuple);

        this.matrix = matrix;

        int length = Math.max(scope[0].getDomain().maxSize(), scope[1]
                .getDomain().maxSize());

        /*
         * No need for last data structure if domain sizes <=
         * MINIMUM_SIZE_FOR_LAST
         */
        if (length > MINIMUM_SIZE_FOR_LAST) {
            last = new int[][] { new int[scope[0].getDomain().maxSize()],
                    new int[scope[1].getDomain().maxSize()] };
        } else {
            last = null;
        }

    }

    public static long getChecks() {
        return checks;
    }

    public static long getPresenceChecks() {
        return presenceChecks;
    }

    public static void clearStats() {
        checks = 0;
        presenceChecks = 0;
    }

    public boolean hasSupport(final int variablePosition, final int index) {
        if (last == null) {
            return hasSupportNR(variablePosition, index);
        }
        return hasSupportR(variablePosition, index);
    }

    private boolean hasSupportR(final int variablePosition, final int index) {
        if (controlResidue(variablePosition, index)) {
            return true;
        }

        final BitVector matrixBV = matrix.getBitVector(variablePosition, index);
        final int intersection = matrixBV
                .intersects(variables[1 - variablePosition].getBitDomain());

        if (intersection >= 0) {
            checks += 1 + intersection;
            last[variablePosition][index] = intersection;
            return true;
        }

        checks += matrixBV.realSize();
        return false;
    }

    private boolean hasSupportNR(final int variablePosition, final int index) {
        checks++;
        return matrix.getBitVector(variablePosition, index).intersects(
                variables[1 - variablePosition].getBitDomain()) >= 0;
    }

    public String toString() {
        return "MatrixManager2D\n" + matrix;
    }

    private boolean controlResidue(final int position, final int index) {
        final int part = last[position][index];
        presenceChecks++;
        return part != -1
                && matrix.getBitVector(position, index).intersects(
                        variables[1 - position].getBitDomain(), part);
    }

    @Override
    public Matrix unshareMatrix() {
        matrix = (Matrix2D) super.unshareMatrix();
        return matrix;
    }
}

File: src/main/scala/cspfj/constraint/extension/MatrixManager2D.scala
package cspfj.constraint.extension;

import cspfj.problem.Variable;
import cspfj.util.BitVector;

object MatrixManager2D {
  var checks = 0;
  var presenceChecks = 0

  def clearStats() {
    checks = 0;
    presenceChecks = 0;
  }

};

final class MatrixManager2D(
  scope: Array[Variable],
  private var matrix2d: Matrix2D,
  shared: Boolean,
  tuple: Array[Int]) extends AbstractMatrixManager(scope, matrix2d, shared, tuple) {

  private val MINIMUM_SIZE_FOR_LAST = 3 * java.lang.Long.SIZE;

  /**
   * No need for last data structure if domain sizes <=
   * MINIMUM_SIZE_FOR_LAST
   */
  private val last = if (scope.map(_.dom.maxSize).max > MINIMUM_SIZE_FOR_LAST) {
    Array(new Array[Int](scope(0).dom.maxSize), new Array[Int](scope(1).dom.maxSize))
  } else {
    null
  }

  def hasSupport(variablePosition: Int, index: Int) = {
    if (last == null) {
      hasSupportNR(variablePosition, index);
    } else
      hasSupportR(variablePosition, index);
  }

  private def hasSupportR(variablePosition: Int, index: Int) = {
    controlResidue(variablePosition, index) || {
      val matrixBV = matrix2d.getBitVector(variablePosition, index);
      val intersection = matrixBV
        .intersects(scope(1 - variablePosition).dom.getBitVector);

      if (intersection >= 0) {
        MatrixManager2D.checks += 1 + intersection;
        last(variablePosition)(index) = intersection;
        true;
      } else {
        MatrixManager2D.checks += matrixBV.realSize;
        false;
      }
    }
  }

  private def hasSupportNR(variablePosition: Int, index: Int) = {
    MatrixManager2D.checks += 1;
    matrix2d.getBitVector(variablePosition, index).intersects(
      scope(1 - variablePosition).dom.getBitVector) >= 0;
  }

  private def controlResidue(position: Int, index: Int) = {
    val part = last(position)(index)
    MatrixManager2D.presenceChecks += 1
    (part != -1 &&
      matrix2d.getBitVector(position, index).intersects(
        scope(1 - position).dom.getBitVector, part))
  }

  override def unshareMatrix() = {
    matrix2d = super.unshareMatrix.asInstanceOf[Matrix2D]
    matrix;
  }
}
